pyqtorch/modules/abstract.py

```diff
@@ -15,21 +16,24 @@ def __init__(self, qubits: ArrayLike, n_qubits: int):
         self.qubits = qubits
         self.n_qubits = n_qubits
 
-    def __mul__(
-        self, other: AbstractGate | pyqtorch.modules.QuantumCircuit
-    ) -> pyqtorch.modules.QuantumCircuit:
+    def __mul__(self, other: AbstractGate | pyq.QuantumCircuit) -> pyq.QuantumCircuit:
         if isinstance(other, AbstractGate):
-            return pyqtorch.modules.QuantumCircuit(
-                max(self.n_qubits, other.n_qubits), [self, other]
-            )
+            ml = torch.nn.ModuleList([self, other])
+            return pyq.QuantumCircuit(max(self.n_qubits, other.n_qubits), ml)
+        elif isinstance(other, pyq.QuantumCircuit):
+            ml = torch.nn.ModuleList([self]) + other.operations
+            return pyq.QuantumCircuit(max(self.n_qubits, other.n_qubits), ml)
+        else:
+            return ValueError(f"Cannot compose {type(self)} with {type(other)}")
 
-        if isinstance(other, pyqtorch.modules.QuantumCircuit):
-            return pyqtorch.modules.QuantumCircuit(
-                max(self.n_qubits, other.n_qubits), [self, *other.operations]
-            )
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, type(self)):
+            return False
 
-        else:
-            return NotImplemented
+        return self.qubits == other.qubits and self.n_qubits == other.n_qubits
+
+    def __hash__(self):
+        return super().__hash__()
```

Comment:
if we don't add this we are getting errors like
```
TypeError: unhashable type: 'X'
```
I am not sure why they are popping up when adding `__eq__` methods. I have no idea how pytorch is implementing this (also could not find out in a reasonable amount of time), but our gates/circuits should strictly speaking not even have been hashable in the first place, because they have the `qubits` field which is a `list` (which is not hashalbe)...